consensus/dolev_strong.go

```diff
@@ -0,0 +1,310 @@
+		return nil
+	}
+	//set the message as the correct one
+	dsci.value = message.Msg.Data
+	log.Info("Message received on node : %v", dsci.ds.publicKey.String())
+	// add our signature on the message
+	err = dsci.ds.signMessageAndAppend(message)
+	if err != nil {
+		return fmt.Errorf("cannot sign message %v error %v", message, err)
+	}
+	//add this node to the nodes already signed
+	publicKeys[dsci.ds.publicKey.String()] = struct{}{}
+	dsci.ds.sendToRemainingNodes(message, publicKeys) // to be sent in round +1
+	return nil
+}
+
+func (dsci *DolevStrongInstance) findAndValidateSignatures(message *pb.ConsensusMessage) (map[string]struct{}, error){
+	publicKeys := make(map[string]struct{})
+	// validate signatures
+	for _, val := range message.Validators {
+		key, ok := crypto.NewPublicKey(val.AuthPubKey)
+		if ok != nil {
+			return nil, errors.New("cannot read validator public key")
+		}
+
+		validSig := dsci.authenticateSignature(key, val.AuthorSign, message.Msg)
+		if !validSig {
+			return nil, fmt.Errorf("invalid signature of %v", val.AuthorSign)
+		}
+		publicKeys[key.String()] = struct{}{}
+	}
+	return publicKeys, nil
+}
+
+// Abort aborts the run of the current agreement protocol
+func (impl *DolevStrongConsensus) Abort() {
+	impl.abortInstance <- struct{}{}
+	impl.abortListening <- struct{}{}
+}
+
+func (impl *DolevStrongConsensus) sendToRemainingNodes(message *pb.ConsensusMessage, foundValidators map[string]struct{}) error {
+	// Sends this message in order to reach consensus
+	data, ok := proto.Marshal(message)
+	if ok != nil {
+		log.Error("could not marshal output message", message)
+	}
+	for key := range impl.activeInstances {
+		//don't send to signed validators
+		if _, ok := foundValidators[key]; ok {
+			continue
+		}
+		_, err := impl.net.SendMessage(data, key)
+		if err != nil {
+			//todo: should we break the loop now?
+			//return err
+		}
+	}
+	return nil
+}
+
+// GetOtherInstancesOutput returns all the outputs agreed in this layer from other instances that ran in the same time as this initiator
+func (impl *DolevStrongConsensus) GetOtherInstancesOutput() map[string][]byte {
+	output := make(map[string][]byte)
+	for key, ds := range impl.activeInstances {
+		output[key] = ds.GetOutput()
+	}
+	return output
+}
+
+func (impl *DolevStrongConsensus) signMessageAndAppend(message *pb.ConsensusMessage) error {
+	//marshall the message so that we could sign the entire byte stream as a whole
+	bytes, err := proto.Marshal(message.Msg)
+	if err != nil {
+		log.Error("cannot marshal message %v", message.Msg)
+		return err
+	}
+	mySignature, err := impl.privateKey.Sign(bytes)
+	if err != nil {
+		return err
+	}
+	validator := pb.Validator{
+		AuthPubKey: impl.publicKey.Bytes(),
+		AuthorSign: mySignature,
+	}
+
+	message.Validators = append(message.Validators, &validator)
+	return nil
+}
+
+//NewCAInstance return a new instance of a dolev strong receiver
+func NewCAInstance(ca *DolevStrongConsensus) CAInstance {
```

Comment:
New* methods should be creating concrete types, it should be named NewDolevStrongInstance